File: day10/src/bin/part2.py
from functools import reduce
import sys

MATCHING = {
    '}': '{',
    ']': '[',
    ')': '(',
    '>': '<',
}

REVMATCHING = {
    '{': '}',
    '[': ']',
    '(': ')',
    '<': '>',
}

POINTS = {
    ')': 1,
    ']': 2,
    '}': 3,
    '>': 4,
}

def solve(lines):
    scores = []
    for line in lines:
        stack = []
        valid = True
        for c in line:
            if c in '{([<':
                stack.append(c)
            if c in '})]>':
                p = stack.pop()
                if MATCHING[c] != p:
                    valid = False
                    break
        if valid:
            missing = ''.join(map(lambda x: REVMATCHING[x], reversed(stack)))
            score = reduce(lambda acc, y: acc*5 + POINTS[y], missing, 0)
            scores.append(score)
            print(f"Missing : {missing}, score : {score}")

    print(sorted(scores)[len(scores)//2])

def main():
    if len(sys.argv) != 2:
        print("Wrong number of args")
        sys.exit(1)

    with open(sys.argv[1]) as fd:
        lines = fd.readlines()

    solve(lines)

if __name__ == "__main__":
    main()

File: day10/src/bin/part2.rs
use std::collections::HashMap;
use std::env;
use std::fs;
use std::path::Path;
use std::process;

fn read_input<P: AsRef<Path>>(filename: P) -> Vec<String> {
    let content = fs::read_to_string(filename).expect("Couldn't read from file");
    content.lines().map(|x| String::from(x)).collect()
}

fn main() {
    let matching = HashMap::from([
        ('}', '{'),
        (']', '['),
        (')', '('),
        ('>', '<'),
    ]);
    let revmatching = HashMap::from([
        ('{', '}'),
        ('[', ']'),
        ('(', ')'),
        ('<', '>'),
    ]);
    let points = HashMap::from([
        (')', 1),
        (']', 2),
        ('}', 3),
        ('>', 4),
    ]);
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Wrong number of args");
        process::exit(1);
    }
    let lines = read_input(&args[1]);
    let mut scores: Vec<i64> = Vec::new();
    for line in lines {
        let mut stack: Vec<char> = Vec::new();
        let mut valid = true;
        for c in line.chars() {
            match c {
                '{'|'('|'['|'<' => stack.push(c),
                '}'|')'|']'|'>' => {
                    let p = stack.pop().unwrap();
                    if p != *matching.get(&c).unwrap() {
                        valid = false;
                        break;
                    }
                },
                _ => (),
            }
        }
        if valid {
            stack.reverse();
            let missing: String = stack.iter().map(|x| revmatching.get(x).unwrap()).collect();
            scores.push(missing.chars().fold(0, |acc, c| acc*5 + points.get(&c).unwrap()));
            println!("Missing {}", missing);
        }
    }
    scores.sort();
    println!("{}", scores[scores.len()/2]);
}
